File: q731/q731.py
class MyCalendarTwo:
    def __init__(self) -> None:
        self.booked = []
        self.overlaps = []

    def book(self, start: int, end: int) -> bool:
        if any(s < end and start < e for s, e in self.overlaps):
            return False
        for s, e in self.booked:
            if s < end and start < e:
                self.overlaps.append((max(s, start), min(e, end)))
        self.booked.append((start, end))
        return True


if __name__ == '__main__':
    s = MyCalendarTwo()
    print(s.book(10, 20))
    print(s.book(50, 60))
    print(s.book(10, 40))
    print(s.book(5, 15))
    print(s.book(5, 10))
    print(s.book(25, 55))

File: q731/q731.go
package main

import "fmt"

func main() {
	m := Constructor()
	fmt.Println(m.Book(10, 20))
	fmt.Println(m.Book(50, 60))
	fmt.Println(m.Book(10, 40))
	fmt.Println(m.Book(5, 15))
	fmt.Println(m.Book(5, 10))
	fmt.Println(m.Book(25, 55))
}

type pair struct {
	start, end int
}

type MyCalendarTwo struct {
	booked, overlaps []pair
}

func Constructor() MyCalendarTwo {
	return MyCalendarTwo{}
}

func (c *MyCalendarTwo) Book(start, end int) bool {
	for _, p := range c.overlaps {
		if p.start < end && start < p.end {
			return false
		}
	}
	for _, p := range c.booked {
		if p.start < end && start < p.end {
			c.overlaps = append(c.overlaps, pair{max(p.start, start), min(p.end, end)})
		}
	}
	c.booked = append(c.booked, pair{start, end})
	return true
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
